tests/tracking/fluent/test_fluent.py

```diff
@@ -1498,7 +1498,9 @@ def test_set_experiment_thread_safety(tmp_path):
     origin_create_experiment = MlflowClient.create_experiment
 
     def patched_create_experiment(self, *args, **kwargs):
-        time.sleep(2)
+        # The sleep is for triggering `mlflow.set_experiment`
```

Comment:
what happens if we ran this test on mlflow 2.17.0?

Reply:
it should fail

Reply:
How does it fail?

Reply:
I have explained it in https://github.com/mlflow/mlflow/pull/13577#discussion_r1820037921

in old version MLflow the following race condition might happen:

when user calls `mlflow.set_experiment` with the same name concurrently from 2 threads or processes, it might result in 2 different experiments are created but with the same experiment ID.

this test will triggers the race condition and fail.

Reply:
Ran this test on branch-2.17. It constantly fails without sleep.

Reply:
> Ran this test on branch-2.17. It constantly fails without sleep.

branch-2.17 has other issues besides the race condition issue. so it might fail in other places. You can paste error message and we can check it.

Reply:
@WeichenXu123 

`assert len(created_exp_ids) == 2` fails because `len(created_exp_ids) == 1`.

Reply:
essentially, the sleep is used for capturing the following race condition:

assuming **thread-A** do:

1. get_object_by_name "X"
2. if (1) get None, create object of name "X", and return the created one.

assuming **thread-B** do:

1. get_object_by_name "X"
2. if (1) get None, create object of name "X", and return the created one.

So the race condition happens if the following sequence occurs:

thread-A-step-1, thread-B-step-1, thread-A-step-2, thread-B-step-2

in this race condition case, thread A and B will both create object with name "X", which is faulty.

But the race condition might not always happen, because step(1) and step(2) might finish quickly, so in most cases, the execution sequence will be like:

thread-A-step-1, thread-A-step-2, thread-B-step-1, thread-B-step-2,
or
thread-B-step-1, thread-B-step-2, thread-A-step-1, thread-A-step-2,

and does not trigger the race condition.

To stably reproduce the race condition,
we need to insert a sleep between step 1 and step 2, then if we start 2 threads in nearly the same time, the 2 threads steps are ensured to be executed interlacedly.

Reply:
Makes sense, thanks for the explanation!